File: backend-java/src/main/java/com/example/modules/orders/internal/OrderService.java
// Business logic — hidden from the outside world.
package com.example.modules.orders.internal;

import java.util.Map;
import java.util.Optional;
import java.util.concurrent.ConcurrentHashMap;

public class OrderService {
    public static class Order {
        public String orderId, status, paymentId, cancelReason;
        public long totalCents, refundedCents;
        public Order(String orderId, long totalCents, String status) {
            this.orderId = orderId; this.totalCents = totalCents; this.status = status;
        }
    }

    private static final Map<String, Order> STORE = new ConcurrentHashMap<>();

    public Order createOrder(String orderId, long totalCents) {
        Order o = new Order(orderId, totalCents, "PENDING");
        STORE.put(orderId, o);
        return o;
    }

    public Optional<Order> getOrder(String orderId) { return Optional.ofNullable(STORE.get(orderId)); }

    public Optional<Order> markPaid(String orderId, String paymentId) {
        Order o = STORE.get(orderId);
        if (o == null) return Optional.empty();
        o.status = "PAID"; o.paymentId = paymentId;
        return Optional.of(o);
    }

    public Optional<Order> markFailed(String orderId, String reason) {
        Order o = STORE.get(orderId);
        if (o == null) return Optional.empty();
        o.status = "CANCELLED"; o.cancelReason = reason;
        return Optional.of(o);
    }

    public Optional<Order> markRefunded(String orderId, long amountCents) {
        Order o = STORE.get(orderId);
        if (o == null || !"PAID".equals(o.status)) return Optional.empty();
        o.status = "REFUNDED"; o.refundedCents = amountCents;
        return Optional.of(o);
    }
}

File: backend-csharp/modules/orders/Internal/OrderService.cs
// Business logic — hidden from the outside world.
using System.Collections.Concurrent;
namespace BackendCsharp.Modules.Orders.Internal;

public class Order
{
    public string OrderId { get; set; } = "";
    public long TotalCents { get; set; }
    public string Status { get; set; } = "PENDING";
    public string? PaymentId { get; set; }
    public string? CancelReason { get; set; }
    public long? RefundedCents { get; set; }
}

public class OrderService
{
    private static readonly ConcurrentDictionary<string, Order> Store = new();

    public Order CreateOrder(string orderId, long totalCents)
    {
        var o = new Order { OrderId = orderId, TotalCents = totalCents, Status = "PENDING" };
        Store[orderId] = o;
        return o;
    }

    public Order? GetOrder(string orderId) => Store.GetValueOrDefault(orderId);

    public Order? MarkPaid(string orderId, string paymentId)
    {
        if (!Store.TryGetValue(orderId, out var o)) return null;
        o.Status = "PAID"; o.PaymentId = paymentId;
        return o;
    }

    public Order? MarkFailed(string orderId, string reason)
    {
        if (!Store.TryGetValue(orderId, out var o)) return null;
        o.Status = "CANCELLED"; o.CancelReason = reason;
        return o;
    }

    public Order? MarkRefunded(string orderId, long amountCents)
    {
        if (!Store.TryGetValue(orderId, out var o) || o.Status != "PAID") return null;
        o.Status = "REFUNDED"; o.RefundedCents = amountCents;
        return o;
    }
}
